TypeSystem/Reflection/schema.h

```diff
@@ -0,0 +1,231 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_TYPE_SYSTEM_REFLECTION_SCHEMA_H
+#define CURRENT_TYPE_SYSTEM_REFLECTION_SCHEMA_H
+
+#include "reflection.h"
+
+#include "../../Bricks/strings/strings.h"
+
+namespace current {
+namespace reflection {
+
+CURRENT_STRUCT(TypeInfo) {
+  CURRENT_FIELD(type_id, uint64_t, 0u);
+  CURRENT_FIELD(included_types, std::vector<uint64_t>);
+};
+
+CURRENT_STRUCT(StructInfo) {
+  CURRENT_FIELD(type_id, uint64_t, 0u);
+  CURRENT_FIELD(name, std::string, "");
+  CURRENT_FIELD(super_type_id, uint64_t, 0u);
+  using pair_id_name = std::pair<uint64_t, std::string>;
+  CURRENT_FIELD(fields, std::vector<pair_id_name>);
+
+  CURRENT_DEFAULT_CONSTRUCTOR(StructInfo) {}
+  CURRENT_CONSTRUCTOR(StructInfo)(const ReflectedTypeImpl* r) {
+    const ReflectedType_Struct* reflected_struct = dynamic_cast<const ReflectedType_Struct*>(r);
+    if (reflected_struct != nullptr) {
+      type_id = static_cast<uint64_t>(reflected_struct->type_id);
+      name = reflected_struct->name;
+      const ReflectedType_Struct* super = reflected_struct->reflected_super;
+      if (super) {
+        super_type_id = static_cast<uint64_t>(super->type_id);
+      }
+      for (const auto& f : reflected_struct->fields) {
+        fields.emplace_back(static_cast<uint64_t>(f.first->type_id), f.second);
+      }
+    }
+  }
+};
+
+CURRENT_STRUCT(SchemaInfo) {
+  using map_typeid_structinfo = std::map<uint64_t, StructInfo>;
+  CURRENT_FIELD(structs, map_typeid_structinfo);
+  using map_typeid_typeinfo = std::map<uint64_t, TypeInfo>;
+  CURRENT_FIELD(types, map_typeid_typeinfo);
+  CURRENT_FIELD(ordered_struct_list, std::vector<uint64_t>);
+};
+
+struct StructSchema {
+  StructSchema() {
+#define CURRENT_DECLARE_PRIMITIVE_TYPE(typeid_index, cpp_type, unused_current_type) \
+  primitive_type_id_name_[TYPEID_BASIC_TYPE + typeid_index] = #cpp_type;
+#include "../primitive_types.dsl.h"
+#undef CURRENT_DECLARE_PRIMITIVE_TYPE
+  }
+
+  StructSchema(const SchemaInfo& schema) : StructSchema() { schema_ = schema; }
```

Comment:
Won't comma notation work here? If no, I'd move the common initialization into a dedicated private member, just to avoid possible confusions.

Alternatively, make a new `PrimitiveTypesSchema` that self-initializes from the DSL in its own constructor, and just use it as a member in `StructSchema`.